python/ray/dashboard/modules/aggregator/aggregator_agent.py

```diff
@@ -230,22 +316,72 @@ def _can_expose_event(self, event) -> bool:
             in self._exposable_event_types
         )
 
-    def _send_events_to_external_service(self, event_batch) -> None:
+    def _create_ray_events_data(self, event_batch, task_events_metadata=None):
+        """
+        Helper method to create RayEventsData from event batch and metadata.
         """
-        Sends a batch of events to the external service via HTTP POST request
+        events_data = events_event_aggregator_service_pb2.RayEventsData()
+        events_data.events.extend(event_batch)
+
+        if task_events_metadata:
+            events_data.task_events_metadata.CopyFrom(task_events_metadata)
+
+        return events_data
+
+    def _send_events_to_gcs(self, event_batch, task_events_metadata=None) -> bool:
+        """
+        Sends a batch of events to GCS via the GCS grpc client.
+        """
+        if not PUBLISH_EVENTS_TO_GCS:
+            return True
+
+        if not event_batch and (
+            not task_events_metadata
+            or len(task_events_metadata.dropped_task_attempts) == 0
+        ):
+            return True
+
+        try:
+            # Create RayEventsData from the event batch
+            events_data = self._create_ray_events_data(
+                event_batch, task_events_metadata
+            )
+
+            request = events_event_aggregator_service_pb2.AddEventsRequest(
+                events_data=events_data
+            )
+            response = self._gcs_event_stub.AddEvents(request)
```

Comment:
Should we specify a reasonable RPC timeout on the stub call so we don't hang forever?